src/realm/dictionary.hpp

```diff
@@ -0,0 +1,189 @@
+/*************************************************************************
+ *
+ * Copyright 2019 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#ifndef REALM_DICTIONARY_HPP
+#define REALM_DICTIONARY_HPP
+
+#include <realm/array.hpp>
+#include <realm/array_mixed.hpp>
+#include <realm/bplustree.hpp>
+
+namespace realm {
+
+class Dictionary : public ArrayParent {
+public:
+    class Iterator;
+    class MixedRef;
+
+    Dictionary() {}
+
+    Dictionary(const ConstObj& obj, ColKey col_key)
+        : m_obj(obj)
+        , m_col_key(col_key)
+        , m_keys(new BPlusTree<Mixed>(obj.get_alloc()))
```

Comment:
Why two different trees for keys and values? Why not a single, but still separate leaf arrays?

Reply:
Those are the nails that are available using the BPlusTree template.